applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+        block_for_each<OrReduction>(mrModelPart.Elements(), [&first_gamma](Element& r_element)
+        {
+            static constexpr double tolerance = 1e-8;
+            const double this_gamma = r_element.GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+            return std::fabs(first_gamma - this_gamma) > tolerance;
+        });
+
+        KRATOS_ERROR_IF(non_uniform_gamma)
+            << "HEAT_CAPACITY_RATIO is not constant in the domain" << std::endl;
+#endif
+    }
+
+    return err_code;
+}
+
+
+const Parameters ShockCapturingEntropyViscosityProcess::GetDefaultParameters() const
+{
+    return Parameters(R"(
+    {
+        "model_part_name" : "",
+        "calculate_nodal_area_at_each_step" : false,
+        "entropy_constant"  : 0.0,
+        "energy_constant"   : 0.0,
+        "artificial_mass_viscosity_Prandtl"   : 0.1,
+        "artificial_conductivity_Prandtl"     : 0.1
+    }
+    )");
+}
+
+
+double ShockCapturingEntropyViscosityProcess::TotalDerivativeUtil::
+Divergence(const Matrix& rShapeFunGradients, const Matrix& rNodalValues)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rShapeFunGradients.size1() != rNodalValues.size1())
+        << "Matrices rShapeFunGradients and rNodalValues must be the same size (Size1: "
+        << rShapeFunGradients.size1() << " != " << rNodalValues.size1() <<")" << std::endl;
+    KRATOS_DEBUG_ERROR_IF(rShapeFunGradients.size2() != rNodalValues.size2())
+        << "Matrices rShapeFunGradients and rNodalValues must be the same size (Size2: "
+        << rShapeFunGradients.size2() << " != " << rNodalValues.size2() <<")" << std::endl;
+
+    double divergence = 0.0;
+    for(std::size_t i=0; i<rShapeFunGradients.size1(); ++i)
+    {
+        for(std::size_t j=0; j<rShapeFunGradients.size2(); ++j)
+        {
+            divergence += rShapeFunGradients(i,j) * rNodalValues(i,j);
+        }
+    }
+    return divergence;
+
+    KRATOS_CATCH("")
+}
+
+
+void ShockCapturingEntropyViscosityProcess::ExecuteInitializeSolutionStep()
+{
+    if(mFirstTimeStep)
+    {
+        UpdateNodalAreaProcess();
+        ComputeNodalEntropies<1>();
+        /* ^ Necessary in order to compute derivative in first step.
+         *   Stored in buffer index 1 to prevent it from being overwritten at the
+         *      end of this same time-step.
+         *   Ideally would be computed at ExecuteInitialize but initial condition
+         *      processes don't run until ExecuteInitializeSolutionStep
+         */
+        mFirstTimeStep = false;
+    }
+}
+
+
+void ShockCapturingEntropyViscosityProcess::ExecuteFinalizeSolutionStep()
+{
+
+    if (mComputeAreasEveryStep)
+    {
+        UpdateNodalAreaProcess();
+    }
+
+    ComputeNodalEntropies();
+    ComputeArtificialMagnitudes();
+}
+
+
+void ShockCapturingEntropyViscosityProcess::UpdateNodalAreaProcess()
+{
+    CalculateNodalAreaProcess<false> nodal_area_process(mrModelPart);
+    nodal_area_process.Execute();
+}
+
+
+template<unsigned int WriteBufferIndex>
```

Comment:
I don't see the point on making this a template. Couldn't it be a function argument?

Reply:
Well the point was to have it known at compile time, but yeah it might be a bit overkill.

Reply:
Fixed